### PR TITLE
Fully implement all InsertBefore/InsertAfter overloads

### PR DESCRIPTION
**Pull request by BruceForstall**
Fixes problem where they wouldn't work with insertionPoint==nullptr.

### CODE REVIEW

**Comment by BruceForstall** [comment]
@pgavlin PTAL

File: src/jit/lir.cpp [hunk 1]
```diff
@@ -709,11 +665,102 @@ void LIR::Range::InsertBefore(GenTree* insertionPoint, GenTree* node1, GenTree*
...
+    assert(node4->gtNext == nullptr);
+    assert(node4->gtPrev == nullptr);
+
+    node1->gtNext = node2;
+    node2->gtNext = node3;
+    node3->gtNext = node4;
+
+    node4->gtPrev = node3;
+    node3->gtPrev = node2;
+    node2->gtPrev = node1;
```
Comment (pgavlin) [inline src/jit/lir.cpp]: I'm finding the batched gtNext/gtPrev updates a bit hard to read. Can we simply pair the updates for each  node? i.e.

```
node1->gtNext = node2;

node2->gtPrev = node1;
node2->gtNext = node3;

node3->gtPrev = node2;
node3->gtNext = node4;

node4->gtPrev = node3;
```

**Comment by pgavlin** [comment]
LGTM

File: src/jit/lir.cpp (hunk 1 above)
Comment (BruceForstall) [inline src/jit/lir.cpp]: I could do that if you find that easier.

Comment (pgavlin) [inline src/jit/lir.cpp]: That would be great; thanks!

Comment (mikedn) [inline src/jit/lir.cpp]: Meh, so much duplication :smile:

``` C++
namespace InsertUtils {
    static void Link(GenTree* first, GenTree* second) {
        first->gtNext = second;
        second->gtPrev = first;
    }

    static GenTree* Chain(GenTree* last) {
        return last;
    }

    template<typename ...TRest>
    static GenTree* Chain(GenTree* first, GenTree* second, TRest... rest) {
        Link(first, second);
        return Chain(second, rest...);
    }

    template<typename ...TRest>
    static GenTree* First(GenTree* first, TRest... rest) {
        return first;
    }

    static void InsertChain(GenTree* after, GenTree* first, GenTree* last, GenTree *before) {
        if (after != nullptr) {
            Link(after, first);
        }

        if (before != nullptr) {
            Link(last, before);
        }
    }
};

template<typename ...TNodes>
void InsertAfter(GenTree* after, TNodes... nodes) {
    GenTree* first = InsertUtils::First(nodes...);
    GenTree* last = InsertUtils::Chain(nodes...);
    GenTree* before = after->gtNext;
    InsertUtils::InsertChain(after, first, last, before);
}

template<typename ...TNodes>
void InsertBefore(GenTree* before, TNodes... nodes) {
    GenTree* first = InsertUtils::First(nodes...);
    GenTree* last = InsertUtils::Chain(nodes...);
    GenTree* after = before->gtPrev;
    InsertUtils::InsertChain(after, first, last, before);
}
```

**Comment by pgavlin** [comment]
@BruceForstall: the variadic template magic above is what I was wondering about the other day. Thanks @mikedn for figuring that out so I don't have to :D

**Comment by BruceForstall** [comment]
@pgavlin updated PR. Changed range-based insert functions to also use common helper code.

**Comment by pgavlin** [comment]
Awesome!

Also, @mikedn: if you want to put together a PR to the LinearBackendIR branch that Bruce and I have been working in with the changes you suggest, we'd be happy to have it.

**Comment by BruceForstall** [comment]
I'm not sure that meets my "easy to read/understand" bar, but it is cool...

**Comment by mikedn** [comment]
@pgavlin I could do that but I think it's best to do such tweaks after LIR changes reach master, that PR is gigantic as it is...

**Comment by pgavlin** [comment]
Fair enough :)

**Pull request merged**